hudi-spark/src/main/scala/org/apache/hudi/HoodieMergeOnReadRDD.scala

```diff
@@ -0,0 +1,227 @@
+        s"file path: ${mergeParquetPartition.split.dataFile.filePath}" +
+        s"log paths: ${mergeParquetPartition.split.logPaths.toString}" +
+        s"hoodie table path: ${mergeParquetPartition.split.tablePath}" +
+        s"spark partition Index: ${mergeParquetPartition.index}")
+    }
+  }
+
+  override protected def getPartitions: Array[Partition] = {
+    hoodieRealtimeFileSplits.zipWithIndex.map(file => HoodieMergeOnReadPartition(file._2, file._1)).toArray
+  }
+
+  private def getConfig(): Configuration = {
+    broadcastedConf.value.get()
+  }
+
+  private def read(partitionedFile: PartitionedFile,
+                   readFileFunction: PartitionedFile => Iterator[Any]): Iterator[InternalRow] = {
+    val fileIterator = readFileFunction(partitionedFile)
+    val rows = fileIterator.flatMap(_ match {
+      case r: InternalRow => Seq(r)
+      case b: ColumnarBatch => b.rowIterator().asScala
+    })
+    rows
+  }
+
+  private def unMergeFileIterator(split: HoodieMergeOnReadFileSplit,
+                                  baseFileIterator: Iterator[InternalRow]): Iterator[InternalRow] =
+    new Iterator[InternalRow] {
+      private val logSchema = getLogAvroSchema(split)
+      private val sparkTypes = SchemaConverters.toSqlType(logSchema).dataType.asInstanceOf[StructType]
+      private val converter = new AvroDeserializer(logSchema, sparkTypes)
+      private val logRecords = scanLog(split, logSchema).getRecords
+      private val logRecordsIterator = logRecords.keySet().iterator().asScala
+
+      override def hasNext: Boolean = {
+        baseFileIterator.hasNext || logRecordsIterator.hasNext
+      }
+
+      override def next(): InternalRow = {
+        if (baseFileIterator.hasNext) {
+          baseFileIterator.next()
+        } else {
+          val curAvrokey = logRecordsIterator.next()
+          val curAvroRecord = logRecords.get(curAvrokey).getData.getInsertValue(logSchema).get()
+          converter.deserialize(curAvroRecord).asInstanceOf[InternalRow]
+        }
+      }
+    }
+
+  private def mergeFileIterator(split: HoodieMergeOnReadFileSplit,
+                                baseFileIterator: Iterator[InternalRow]): Iterator[InternalRow] =
+    new Iterator[InternalRow] {
+      private val avroSchema = getLogAvroSchema(split)
+      private val sparkSchema = SchemaConverters.toSqlType(avroSchema).dataType.asInstanceOf[StructType]
+      private val avroToRowConverter = new AvroDeserializer(avroSchema, sparkSchema)
+      private val rowToAvroConverter = new AvroSerializer(sparkSchema, avroSchema, false)
+      private val logRecords = scanLog(split, avroSchema).getRecords
+      private val logRecordToRead = logRecords.keySet()
+
+      private var baseFileFinished = false
+      private var logRecordsIterator: Iterator[String] = _
+      private var recordToLoad: InternalRow = _
+
+      @scala.annotation.tailrec
+      override def hasNext: Boolean = {
+        if (baseFileIterator.hasNext) {
```

Comment:
- Instead of looping over `baseFileIterator` and performing the check whether that key exists in `logRecords`, will it be more efficient to do it the other way round. Loop over `logRecords` and perform merge. In the end append all the remaining base file rows.

- Also is this a good practice to perform the actual fetching in `hasNext` function, instead of `next` ?

Reply:
Actually on further thought the first point may not be possible, since if we iterator over log records it will be difficult to find the corresponding base parquet record using record key (for merging).

Reply:
- The log scanner scans all the log files in one batch and handled the merging internally. The output is a hashmap we can use directly. This `logRecordsIterator` is just looping through the hashmap and doesn't load the row one by one like the `daseFileIterator`.

- This is a little bit tricky. If the hasNext return true, but next() doesn't return a value, Spark will throw an exception. In our logic, we don't know `hasNext` will be true of false until we find the qualified record to read. For example, 100 records in base file and 100 delete records in the log file. We will read 0 row and hasNext should return false in the first call, but we have iterated through the whole base file already. There is a test case for this example.